### DISCUSSION TITLE
Cli for build independent usage?

### DISCUSSION THREAD

**Discussion by maxandersen** [openrewrite/rewrite#834] [Ideas] [answered]
Any current work or ideas around having a openrewrite cli one Can Use to apply rewrite rules no matter what build system is in use ?

**Comment by jkschneider** [comment]
Hi @maxandersen! It's technically possible, but what is tough is that to properly type attribute all Java ASTs you have to determine what the binary and project dependencies are for each project + source set in a repository. And determine what those projects and source sets are in the first place. These are of course precisely what build tools are responsible for, so the easiest integration is certainly there. For Maven projects in particular we do have `MavenProjectParser` in `rewrite-maven` which more or less encapsulates these tasks.

For operating _en masse_ across lots of repos (outside of the build), we produce a serialized AST artifact and add that to [moderne.io](https://moderne.io), where we have the whole catalog of recipes that can run across these ASTs and produce PRs and commits and such. We do have a public instance of this that can consume any public repos you'd like -- just have a bit of a waitlist to onboard folks as we continue to scale it out. Feel free to request a beta invite there if you'd like.

**Reply by maxandersen** [reply]
Yes, I get the value in hooking in close with the build system but for some cases its nice to try run on projects without having to touch the build files or what is in my interest being able to use the rewrite rules on jbang.dev based projects. For that having a cli to call with explicit source / lib deps would be really interesting.

i'll take a look at `MavenProjectParser` and maven/gradle plugins and see if I can extract the basics needed to run rewrite programmatically....or is that documented somewhere ?

**Comment by maxandersen** [comment]
Actually managed to make one based on maven plugin (gradles nested classloader code was discouraging  :) 

try out https://github.com/maxandersen/rewrite-jbang :)

**Reply by albertocavalcante** [reply]
Thanks for creating this tool @maxandersen! It's really useful.

I prefer this approach than using the Moderne CLI. It also made it simpler to run with Bazel (as a standalone tool).

**Comment by timtebeek** [comment, marked as answer]
Hi @maxandersen ! Appreciate you're looking to make it easier to run OpenRewrite recipes, and the work you did on `rewrite-jbang` to that end so far.

The good news is we're now also looking to do the same, [as announced publicly](https://www.intelcapital.com/moderne-the-antidote-to-software-maintenance-toil).
So keep an eye out until the day we're ready to share those efforts with you!

**Reply by maxandersen** [reply]
Coming up to the 1 year anniversary. Is there some news around a cli to run? :)

**Reply by timtebeek** [reply]
Hi @maxandersen ; Yes absolutely! We've developed [the Moderne CLI](https://docs.moderne.io/user-documentation/getting-started/cli-intro) over the past year (time flies!), and that's now a perfect companion to [the Moderne platform](https://app.moderne.io/) and [the Moderne IntelliJ plugin](https://plugins.jetbrains.com/plugin/17565-moderne). You'll need a platform token for the CLI, but after that you can run recipes, produce data tables and visualizations locally; reusing the serialized LSTs for Quarkus projects available through the platform. Happy to help you through any questions you might have; feel free to reach out however is convenient.

**Reply by maxandersen** [reply]
moderne cli is cool but its not a cli for running openrewrite rules locally if it needs a moderne app token?

**Reply by timtebeek** [reply]
Recipes do run locally(!), but you can (optionally) use the connection to the Moderne platform to download LSTs that we serialize over night, which could be a large time saver for Quarkus and associated repositories, especially when running across multiple repositories.

We've clarified the [terms around using the Moderne CLI](https://docs.moderne.io/administrator-documentation/licensing) to make it clear what is allowed to specific groups. In your case you're free to use the CLI on OSS projects, but you need a token for compliance and to pull down LSTs and recipes.

Hope that helps!
